lib/bootboot.rb

```diff
@@ -20,13 +20,17 @@ def env_previous
       env_prefix + '_PREVIOUS'
     end
 
+    def load
+      return if Bundler::Plugin.instance_variable_get(:@loaded_plugin_names).include?('bootboot')
```

Comment:
This is reaching private API of `Bundler::Plugin`. Can we avoid this?

Reply:
All of the methods on `Bundler::Plugin` that you might want to use to query whether the plugin was loaded like `Plugin.command?('bootboot')` fail because they query the `Plugin::Index` (which isn't cleared by `register_plugin`) instead of the instance variables of `Plugin` (which are cleared). I think I would actually consider this a bug in Bundler's plugin API since if you follow the instructions on their docs, you will have the same problem, they use a command as an example, but if you used a hook, the hook would not fire if the plugin wasn't already installed. There is only one other way I can think of to fix this issue, if it seems less coupled to internals, I'm happy to change it:

`Plugin.load_plugin` is only ever called if `Plugin.installed?('bootboot')` is true. So, using that would look something like:

```ruby
# plugins.rb
if Plugin.installed?("bootboot")
  require "bootboot"
else
  load File.expand_path("lib/bootboot.rb", __dir__), true
end
```

The main problem with this approach is that it needs to use `load path, true` to isolate the global namespace so that `Bootboot::GEMFILE` isn't defined multiple times for instance, but that doesn't seem to play well with the nested `require` statements so it seemed kind of messy to try to figure that out. For instance, if you `require "bootboot/gemfile_auto_sync"`, that defines `::Bootboot::GemfileAutoSync`, but then back in `bootboot.rb` calling `Bootboot::GemfileAutoSync.new.setup` doesn't working because the constant lookup is actually `<#Module anonymous>::Bootboot::GemfileAutoSync`. I'm sure I could figure it out though, if that would be a better approach.